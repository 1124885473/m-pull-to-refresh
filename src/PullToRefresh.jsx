import React, { PropTypes } from 'react';
import initWebPullToRefresh from './wptr.1.1';
import Hammer from 'react-hammerjs';

const emptyEvents = {
  onPanStart: undefined,
  onPan: undefined,
  onPanEnd: undefined,
};

export default class PullToRefresh extends React.Component {
  static propTypes = {
    loadingFunction: PropTypes.func.isRequired,
    icon: PropTypes.element,
    prefixCls: PropTypes.string,
    loading: PropTypes.element,
    disabled: PropTypes.bool,
    className: PropTypes.string,
    style: PropTypes.object,
    contentClassName: PropTypes.string,
    contentStyle: PropTypes.object,
    distanceToRefresh: PropTypes.number,
    resistance: PropTypes.number,
    children: PropTypes.any,
  };

  static defaultProps = {
    prefixCls: 'rmc-pull-to-refresh',
  };

  componentWillMount() {
    this.webPullToRefresh = initWebPullToRefresh();
  }
  componentDidMount() {
    const { props, refs } = this;
    this.webPullToRefresh.init({
      contentEl: refs.content,
      containerEl: refs.container,
      ptrEl: refs.ptr,
      prefixCls: props.prefixCls,
      distanceToRefresh: props.distanceToRefresh,
      loadingFunction: props.loadingFunction,
      resistance: props.resistance,
    });
  }

  render() {
    const {
      prefixCls, children, icon, loading, disabled, className = '',
      style, contentStyle, contentClassName = '',
    } = this.props;
    const events = disabled ? emptyEvents : this.webPullToRefresh.events;
    return (
      <div className={`${className} ${prefixCls}`} style={style} ref="container">
        <div ref="ptr" className={`${prefixCls}-ptr`}>
          <div className={`${prefixCls}-ptr-icon`}>{icon}</div>
          <div className={`${prefixCls}-ptr-loading`}>{loading}</div>
        </div>
        <Hammer vertical {...events}>
          <div
            ref="content"
            className={`${prefixCls}-content ${contentClassName}`}
            style={contentStyle}
          >
            {children}
          </div>
        </Hammer>
      </div>
    );
  }
}
