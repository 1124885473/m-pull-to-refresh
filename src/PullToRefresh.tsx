import React from 'react';
import ReactDOM from 'react-dom';
import classNames from 'classnames';
import PropsType from './PropsType';

export class StateType {
  pullUp: any;
  isTouching: boolean;
}

function setTransform(nodeStyle: any, value: any) {
  nodeStyle.transform = value;
  nodeStyle.webkitTransform = value;
  nodeStyle.MozTransform = value;
}

export default class PullToRefresh extends React.Component<PropsType, StateType> {
  static defaultProps = {
    prefixCls: 'rmc-pull-to-refresh',
    direction: 'down',
    distanceToRefresh: 25,
  } as PropsType;

  // https://github.com/yiminghe/zscroller/blob/2d97973287135745818a0537712235a39a6a62a1/src/Scroller.js#L355
  // states: `activate` / `deactivate` / `release` / `finish`
  state = {
    pullUp: '',
    isTouching: false,
  };

  pullUpStats: any = {
    activate: 'activate',
    deactivate: 'deactivate',
    release: 'release',
    finish: 'finish',
  };

  pullUpDisplay: any = {
    activate: '释放刷新',
    deactivate: '上拉 ↑',
    release: '加载中...',
    finish: '完成刷新',
  };

  containerRef: any;
  contentRef: any;
  _to: any;
  _pullUpScreenY: any;
  _pullUpStartScreenY: any;
  _pullUpLastScreenY: any;
  _pullUpTimer: any;

  componentDidUpdate(prevProps: any) {
    if (prevProps === this.props) {
      return;
    }
    const preRefreshing = prevProps.refreshing;
    const nowRefreshing = this.props.refreshing;
    if (preRefreshing && !nowRefreshing && !this._pullUpTimer) {
      this.pullUpFinish();
    } else if (!preRefreshing && nowRefreshing) {
      // this.triggerRefresh();
    }
  }

  componentDidMount() {
    if (this.props.getScrollContainer) {
      setTimeout(() => {
        this.initPullUp(this.props.getScrollContainer());
      });
    } else {
      this.initPullUp(ReactDOM.findDOMNode(this.containerRef));
    }
  }

  componentWillUnmount() {
    if (this.props.getScrollContainer) {
      this.destroyPullUp(this.props.getScrollContainer());
    } else {
      this.destroyPullUp(ReactDOM.findDOMNode(this.containerRef));
    }
  }

  genEvtHandler = (ele: any) => {
    return {
      touchstart: this.onTouchStart.bind(this, ele),
      touchmove: this.onTouchMove.bind(this, ele),
      touchend: this.onTouchEnd.bind(this, ele),
      touchcancel: this.onTouchEnd.bind(this, ele),
    };
  }

  initPullUp = (ele: any) => {
    this._to = this.genEvtHandler(ele);
    Object.keys(this._to).forEach(key => {
      ele.addEventListener(key, this._to[key]);
    });
    if (this.props.refreshing) {
      this.setState({ pullUp: this.pullUpStats.release }, () => {
        this._pullUpLastScreenY = - this.props.distanceToRefresh - 1;
        setTransform(this.contentRef.style,
          `translate3d(0px,${this._pullUpLastScreenY}px,0)`);
      });
    }
  }

  destroyPullUp = (ele: any) => {
    Object.keys(this._to).forEach(key => {
      ele.removeEventListener(key, this._to[key]);
    });
  }

  onTouchStart = (_ele: any, e: any) => {
    this._pullUpScreenY = this._pullUpStartScreenY = e.touches[0].screenY;
    // 一开始 refreshing 为 true 时 this._pullUpLastScreenY 有值
    this._pullUpLastScreenY = this._pullUpLastScreenY || 0;
    this.setState({ isTouching: true });
  }

  onTouchMove = (ele: any, e: any) => {

    // 使用 pageY 对比有问题
    const _screenY = e.touches[0].screenY;
    if (this._pullUpStartScreenY - _screenY > 0) {
      // console.log('is pull up', _screenY);

      let isReachBottom;
      if (this.props.getScrollContainer) {
        // In chrome61 `document.body.scrollTop` is invalid, here `ele === document.body`
        const scrollNode = document.scrollingElement ? document.scrollingElement : ele;
        isReachBottom = ele.scrollHeight - scrollNode.scrollTop <= window.innerHeight;
        // console.log(ele.scrollHeight, scrollNode.scrollTop, window.innerHeight);
      } else {
        isReachBottom = ele.scrollHeight - ele.scrollTop === ele.clientHeight;
      }
      if (isReachBottom) {
        const _diff = Math.round(_screenY - this._pullUpScreenY);
        this._pullUpScreenY = _screenY;
        this._pullUpLastScreenY += _diff;

        setTransform(this.contentRef.style,
          `translate3d(0px,${this._pullUpLastScreenY}px,0)`);

        if (Math.abs(this._pullUpLastScreenY) < this.props.distanceToRefresh) {
          if (this.state.pullUp !== this.pullUpStats.deactivate) {
            // console.log('back to the distance');
            this.setState({ pullUp: this.pullUpStats.deactivate });
          }
        } else {
          if (this.state.pullUp === this.pullUpStats.deactivate) {
            // console.log('reach to the distance');
            this.setState({ pullUp: this.pullUpStats.activate });
          }
        }
      }
    }
  }

  onTouchEnd = () => {
    this.setState({ isTouching: false });
    if (this.state.pullUp === this.pullUpStats.deactivate) {
      this.pullUpFinish();
    } else if (this.state.pullUp === this.pullUpStats.activate) {
      this.setState({ pullUp: this.pullUpStats.release });
      this._pullUpTimer = setTimeout(() => {
        if (!this.props.refreshing) {
          this.pullUpFinish();
        }
        this._pullUpTimer = undefined;
      }, 1000);
      this.props.onRefresh();
    }
  }

  pullUpFinish = () => {
    this._pullUpLastScreenY = 0;
    setTransform(this.contentRef.style, `translate3d(0px,0px,0)`);
    if (this.state.pullUp === this.pullUpStats.release) {
      this.setState({ pullUp: this.pullUpStats.finish });
    }
  }

  // triggerRefresh = () => {}

  render() {
    let {
      className, prefixCls, children, getScrollContainer,
      direction, onRefresh, renderer, distanceToRefresh, ...restProps,
    } = this.props;

    // delete restProps.refreshing

    const renderRefresh = () => {
      let defaultRenderer = this.pullUpDisplay.deactivate;
      switch (this.state.pullUp) {
        case 'activate':
        case 'deactivate':
        case 'release':
        case 'finish':
          defaultRenderer = this.pullUpDisplay[this.state.pullUp];
        default:
      }
      const cls = classNames(`${prefixCls}-content`,
        getScrollContainer && `${prefixCls}-${direction}`,
        !this.state.isTouching && this.state.pullUp && `${prefixCls}-transition`,
      );
      return (
        <div className={`${prefixCls}-content-wrapper`}>
          <div className={cls} ref={el => this.contentRef = el}>
            {children}
            <div className={`${prefixCls}-indicator`}>
              {renderer ? renderer(this.state.pullUp) : defaultRenderer}
            </div>
          </div>
        </div>
      );
    };

    if (getScrollContainer) {
      return renderRefresh();
    }
    return (
      <div
        ref={el => this.containerRef = el}
        className={classNames(className, prefixCls, `${prefixCls}-${direction}`)}
        {...restProps}
      >
        {renderRefresh()}
      </div>
    );
  }
}
