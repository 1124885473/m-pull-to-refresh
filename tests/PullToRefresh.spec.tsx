import * as React from 'react';
import * as Enzyme from 'enzyme';
import { render } from 'enzyme';
import renderToJson from 'enzyme-to-json';
import PullToRefresh from '../src';

// Adapter 类似 classnames 的导出，这里为了方便直接用 commonjs
// see https://github.com/JedWatson/classnames/pull/167
const Adapter = require('enzyme-adapter-react-16');

Enzyme.configure({ adapter: new Adapter() });

describe('basic', () => {
  it('base.', () => {
    const wrapper = render(
      <PullToRefresh>
        <div style={{ height: 50 }}>Item 1</div>
      </PullToRefresh>,
    );
    expect(renderToJson(wrapper)).toMatchSnapshot();
  });
});
