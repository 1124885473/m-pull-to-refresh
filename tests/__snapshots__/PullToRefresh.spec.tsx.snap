// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic base. 1`] = `
<div
  class="rmc-pull-to-refresh rmc-pull-to-refresh-down"
>
  <div
    class="rmc-pull-to-refresh-content-wrapper"
  >
    <div
      class="rmc-pull-to-refresh-content rmc-pull-to-refresh-transition"
    >
      <div
        class="rmc-pull-to-refresh-indicator"
      >
        pull
      </div>
      <div>
        <div
          style="height:50px"
        >
          Item 1
        </div>
      </div>
    </div>
  </div>
</div>
`;
